internal/checkcontroller/reconcile.go

```diff
@@ -0,0 +1,518 @@
+			errs = append(errs, fmt.Errorf("error rebooting node: %w", err))
+		}
+	}
+
+	return unschedulableNodes, errs
+}
+
+// checkIfNodeNeedsReboot checks if the node with the given name needs to be rebooted.
+func (r *CheckControllerReconciler) checkIfNodeNeedsReboot(ctx context.Context, node *corev1.Node) (bool, error) {
+	logger := log.FromContext(ctx).WithName("CheckIfNodeNeedsReboot").WithValues("nodeName", node.Name)
+
+	conditionExist, err := r.CheckNodeCondition(ctx, node, SlurmNodeCondition, corev1.ConditionTrue)
+	if err != nil {
+		return false, fmt.Errorf("error checking node condition: %w", err)
+	}
+	if conditionExist {
+		logger.V(1).Info("Node condition exists")
+		return false, nil
+	}
+
+	logger.V(1).Info("Checking if node is drained")
+	if !r.IsNodeDrained(node) {
+		logger.V(1).Info("Node is not drained")
+		return false, nil
+	}
+
+	return false, nil
+}
+
+// rebootNode reboots the node with the given name in the given namespace.
+func (r *CheckControllerReconciler) rebootNode(ctx context.Context, node *corev1.Node, workerName, namespace string) error {
+	logger := log.FromContext(ctx).WithName("RebootNode").WithValues("workerName", workerName, "nodeName", node.Name)
+
+	if err := r.DrainNodeIfNeeded(ctx, node); err != nil {
+		return err
+	}
+
+	if err := r.CreateRebooterPodIfNeeded(ctx, workerName, node.Name, namespace); err != nil {
+		return err
+	}
+
+	if err := r.SetNodeConditionIfNotExists(ctx, node, SlurmNodeCondition, corev1.ConditionTrue, "Node is being rebooted"); err != nil {
+		return err
+	}
+
+	logger.V(1).Info("Node condition set")
+	return nil
+}
+
+// DrainNodeIfNeeded drains the node with the given name if it is not already drained.
+func (r *CheckControllerReconciler) DrainNodeIfNeeded(ctx context.Context, node *corev1.Node) error {
+	logger := log.FromContext(ctx).WithName("DrainNodeIfNeeded").WithValues("nodeName", node.Name)
+
+	if r.IsNodeDrained(node) {
+		logger.V(1).Info("Node is already drained")
+		return nil
+	}
+
+	logger.V(1).Info("Draining node")
+	if err := r.DrainNode(ctx, node); err != nil {
+		return fmt.Errorf("failed to drain node: %w", err)
+	}
+
+	logger.V(1).Info("Marking node as unschedulable")
+	node.Spec.Unschedulable = true
+	if err := r.Update(ctx, node); err != nil {
+		return fmt.Errorf("failed to update nodeto unschedulable: %w", err)
+	}
+
+	logger.V(1).Info("Node drained and marked as unschedulable")
+	return nil
+}
+
+// CreateRebooterPodIfNeeded creates a rebooter pod for the node with the given name if it does not already exist.
+func (r *CheckControllerReconciler) CreateRebooterPodIfNeeded(ctx context.Context, workerName, nodeName, namespace string) error {
+	logger := log.FromContext(ctx).WithName("CreateRebooterPodIfNeeded").WithValues("workerName", workerName, "nodeName", nodeName)
+
+	logger.V(1).Info("Checking for existing rebooter pod")
+	existingPod, err := r.GetRebooterPod(ctx, nodeName, namespace)
+	if err == nil && existingPod != nil {
```

Comment:
what about case `err != nil && !errors.Is(ErrNotFound, err)`?
I mean, you tried to `GetRebooterPod`, it returned an error, but rebooter pod already exists